File: testFiles/NPB3.4.2/NPB3.4-MPI/common/randdp.f90
!---------------------------------------------------------------------
!---------------------------------------------------------------------

      double precision function randlc (x, a)

!---------------------------------------------------------------------
!---------------------------------------------------------------------

!---------------------------------------------------------------------
!
!   This routine returns a uniform pseudorandom double precision number in the
!   range (0, 1) by using the linear congruential generator
!
!   x_{k+1} = a x_k  (mod 2^46)
!
!   where 0 < x_k < 2^46 and 0 < a < 2^46.  This scheme generates 2^44 numbers
!   before repeating.  The argument A is the same as 'a' in the above formula,
!   and X is the same as x_0.  A and X must be odd double precision integers
!   in the range (1, 2^46).  The returned value RANDLC is normalized to be
!   between 0 and 1, i.e. RANDLC = 2^(-46) * x_1.  X is updated to contain
!   the new seed x_1, so that subsequent calls to RANDLC using the same
!   arguments will generate a continuous sequence.
!
!   This routine should produce the same results on any computer with at least
!   48 mantissa bits in double precision floating point data.  On 64 bit
!   systems, double precision should be disabled.
!
!   David H. Bailey     October 26, 1990
!
!---------------------------------------------------------------------

      implicit none

      double precision r23,r46,t23,t46,a,x,t1,t2,t3,t4,a1,a2,x1,x2,z
      parameter (r23 = 0.5d0 ** 23, r46 = r23 ** 2, t23 = 2.d0 ** 23,  &
     &  t46 = t23 ** 2)

!---------------------------------------------------------------------
!   Break A into two parts such that A = 2^23 * A1 + A2.
!---------------------------------------------------------------------
      t1 = r23 * a
      a1 = int (t1)
      a2 = a - t23 * a1

!---------------------------------------------------------------------
!   Break X into two parts such that X = 2^23 * X1 + X2, compute
!   Z = A1 * X2 + A2 * X1  (mod 2^23), and then
!   X = 2^23 * Z + A2 * X2  (mod 2^46).
!---------------------------------------------------------------------
      t1 = r23 * x
      x1 = int (t1)
      x2 = x - t23 * x1
      t1 = a1 * x2 + a2 * x1
      t2 = int (r23 * t1)
      z = t1 - t23 * t2
      t3 = t23 * z + a2 * x2
      t4 = int (r46 * t3)
      x = t3 - t46 * t4
      randlc = r46 * x

      return
      end




!---------------------------------------------------------------------
!---------------------------------------------------------------------

      subroutine vranlc (n, x, a, y)

!---------------------------------------------------------------------
!---------------------------------------------------------------------

!---------------------------------------------------------------------
!
!   This routine generates N uniform pseudorandom double precision numbers in
!   the range (0, 1) by using the linear congruential generator
!
!   x_{k+1} = a x_k  (mod 2^46)
!
!   where 0 < x_k < 2^46 and 0 < a < 2^46.  This scheme generates 2^44 numbers
!   before repeating.  The argument A is the same as 'a' in the above formula,
!   and X is the same as x_0.  A and X must be odd double precision integers
!   in the range (1, 2^46).  The N results are placed in Y and are normalized
!   to be between 0 and 1.  X is updated to contain the new seed, so that
!   subsequent calls to VRANLC using the same arguments will generate a
!   continuous sequence.  If N is zero, only initialization is performed, and
!   the variables X, A and Y are ignored.
!
!   This routine is the standard version designed for scalar or RISC systems.
!   However, it should produce the same results on any single processor
!   computer with at least 48 mantissa bits in double precision floating point
!   data.  On 64 bit systems, double precision should be disabled.
!
!---------------------------------------------------------------------

      implicit none

      integer i,n
      double precision y,r23,r46,t23,t46,a,x,t1,t2,t3,t4,a1,a2,x1,x2,z
      dimension y(*)
      parameter (r23 = 0.5d0 ** 23, r46 = r23 ** 2, t23 = 2.d0 ** 23,  &
     &  t46 = t23 ** 2)


!---------------------------------------------------------------------
!   Break A into two parts such that A = 2^23 * A1 + A2.
!---------------------------------------------------------------------
      t1 = r23 * a
      a1 = int (t1)
      a2 = a - t23 * a1

!---------------------------------------------------------------------
!   Generate N results.   This loop is not vectorizable.
!---------------------------------------------------------------------
      do i = 1, n

!---------------------------------------------------------------------
!   Break X into two parts such that X = 2^23 * X1 + X2, compute
!   Z = A1 * X2 + A2 * X1  (mod 2^23), and then
!   X = 2^23 * Z + A2 * X2  (mod 2^46).
!---------------------------------------------------------------------
        t1 = r23 * x
        x1 = int (t1)
        x2 = x - t23 * x1
        t1 = a1 * x2 + a2 * x1
        t2 = int (r23 * t1)
        z = t1 - t23 * t2
        t3 = t23 * z + a2 * x2
        t4 = int (r46 * t3)
        x = t3 - t46 * t4
        y(i) = r46 * x
      enddo

      return
      end

File: testFiles/NPB3.4.2/NPB3.4-MPI/common/randdp.c
//---------------------------------------------------------------------
//   This function is C verson of random number generator randdp.f 
//---------------------------------------------------------------------

double	randlc(X, A)
double *X;
double *A;
{
      static int        KS=0;
      static double	R23, R46, T23, T46;
      double		T1, T2, T3, T4;
      double		A1;
      double		A2;
      double		X1;
      double		X2;
      double		Z;
      int     		i, j;

      if (KS == 0) 
      {
        R23 = 1.0;
        R46 = 1.0;
        T23 = 1.0;
        T46 = 1.0;
    
        for (i=1; i<=23; i++)
        {
          R23 = 0.50 * R23;
          T23 = 2.0 * T23;
        }
        for (i=1; i<=46; i++)
        {
          R46 = 0.50 * R46;
          T46 = 2.0 * T46;
        }
        KS = 1;
      }

/*  Break A into two parts such that A = 2^23 * A1 + A2 and set X = N.  */

      T1 = R23 * *A;
      j  = T1;
      A1 = j;
      A2 = *A - T23 * A1;

/*  Break X into two parts such that X = 2^23 * X1 + X2, compute
    Z = A1 * X2 + A2 * X1  (mod 2^23), and then
    X = 2^23 * Z + A2 * X2  (mod 2^46).                            */

      T1 = R23 * *X;
      j  = T1;
      X1 = j;
      X2 = *X - T23 * X1;
      T1 = A1 * X2 + A2 * X1;
      
      j  = R23 * T1;
      T2 = j;
      Z = T1 - T23 * T2;
      T3 = T23 * Z + A2 * X2;
      j  = R46 * T3;
      T4 = j;
      *X = T3 - T46 * T4;
      return(R46 * *X);
} 
